File: python/simulator.py
import socket
import random
import time

def generateRandomData():
    dataType = random.choice([1, 2])
    protocol = random.choice([66, 67, 68])
    utcTime = time.strftime("%y%m%d%H%M%S", time.gmtime())
    status = random.choice([0, 1])
    identifier = ''.join(random.choice('0123456789ABCDEF') for _ in range(3))

    return dataType, protocol, utcTime, status, identifier

def sendDataToServer(socket, udpIp, udpPort):
    try:
        dataType, protocol, utcTime, status, identifier = generateRandomData()

        # Verifique se os dados são válidos antes de criar o pacote
        if not (1 <= dataType <= 2) or not (66 <= protocol <= 68):
            raise ValueError("Tipo de dados ou protocolo inválido")

        packet = f">DATA{dataType},{protocol},{utcTime},{status};ID={identifier}<"
        print("\nEnviando dado:")
        print("\n" + packet)
        socket.sendto(packet.encode('utf-8'), (udpIp, udpPort))
    except ValueError as e:
        print(f"Erro ao gerar dados: {e}")
        # Aqui você pode adicionar qualquer lógica adicional ou tratamento necessário.


def simulator(udpIp, udpPort):
    clientSocket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    print("Simulador iniciado, dados serão enviados a cada 5 segundos. \nCtrl + C para encerrar o simulador")

    try:
        while True:
            sendDataToServer(clientSocket, udpIp, udpPort)
            time.sleep(5)
    except KeyboardInterrupt:
        print("Simulador interrompido.")
    finally:
        closeSimulator(clientSocket)
        
def closeSimulator(clientSocket):
    clientSocket.close()

if __name__ == "__main__":
    udpIp = "127.0.0.1"
    udpPort = 12345

    simulator(udpIp, udpPort)

File: node/simulator.js
const dgram = require('dgram');
const moment = require('moment');

function getRandomIntInclusive(min, max) {
    min = Math.ceil(min);
    max = Math.floor(max);
    return Math.floor(Math.random() * (max - min + 1) + min);
}

function generateRandomData() {
    const dataType = getRandomIntInclusive(1, 2);
    const protocol = getRandomIntInclusive(66, 68);
    const utcTime = moment().utc().format("YYMMDDHHmmss");
    const status = getRandomIntInclusive(0, 1);
    const identifier = Math.random().toString(16).substr(2, 3).toUpperCase();

    return { dataType, protocol, utcTime, status, identifier };
}

function sendDataToServer(socket, udpIp, udpPort) {
    const { dataType, protocol, utcTime, status, identifier } = generateRandomData();

    const packet = `>DATA${dataType},${protocol},${utcTime},${status};ID=${identifier}<`;
    console.log('\nEnviando dado:\n', packet);

    socket.send(Buffer.from(packet), udpPort, udpIp, (err) => {
        if (err) {
            console.error('Erro ao enviar o pacote:', err);
        }
    });
}

function simulator(udpIp, udpPort) {
    const clientSocket = dgram.createSocket('udp4');
    console.log('Simulador iniciado, dados serão enviados a cada 5 segundos. \nCtrl + C para encerrar o simulador');
    setTimeout(() => {
        const interval = setInterval(() => {
            sendDataToServer(clientSocket, udpIp, udpPort);
        }, 5000);

        process.on('SIGINT', () => {
            clearInterval(interval);
            console.log('Simulador interrompido.');
            clientSocket.close();
            process.exit();
        });
    }, 3000);
}

const udpIp = '127.0.0.1';
const udpPort = 12345;

simulator(udpIp, udpPort);
